service/storage_service.cc

```diff
@@ -2756,15 +2757,51 @@ storage_service::prepare_replacement_info(std::unordered_set<gms::inet_address>
     if (!replace_host_id) {
         replace_host_id = _gossiper.get_host_id(replace_address);
     }
-    slogger.info("Host {}/{} is replacing {}/{}", get_token_metadata().get_my_id(), get_broadcast_address(), replace_host_id, replace_address);
-    co_await _gossiper.reset_endpoint_state_map();
 
-    co_return replacement_info {
+    auto ri = replacement_info {
         .tokens = std::move(tokens),
         .dc_rack = std::move(dc_rack),
         .host_id = std::move(replace_host_id),
         .address = replace_address,
     };
+
+    auto ignore_hioas = parse_node_list(_db.local().get_config().ignore_dead_nodes_for_replace());
+
+    for (auto hioa : ignore_hioas) {
+        gms::inet_address addr;
+        // Resolve both host_id and endpoint
+        if (!hioa.has_endpoint()) {
+            auto res = _gossiper.get_nodes_with_host_id(hioa.id);
+            if (res.size() == 0) {
+                throw std::runtime_error(::format("Could not find ignored node with host_id {}", hioa.id));
+            } else if (res.size() > 1) {
+                throw std::runtime_error(::format("Found multiple nodes to ignore with host_id {}: {}", hioa.id, res));
+            }
+            hioa.endpoint = *res.begin();
+        }
+        auto esp = _gossiper.get_endpoint_state_ptr(hioa.endpoint);
+        if (!esp) {
+            throw std::runtime_error(::format("Ignore node {}/{} has no endpoint state", hioa.id, hioa.endpoint));
+        }
+        if (!hioa.has_host_id()) {
+            hioa.id = esp->get_host_id();
+            if (!hioa.id) {
+                throw std::runtime_error(::format("Could not find host_id for ignored node {}", hioa.endpoint));
+            }
+        }
+        auto opt_dc_rack = esp->get_dc_rack();
+        tmptr->update_topology(hioa.id, opt_dc_rack);
+        if (hioa.id) {
+            tmptr->update_host_id(hioa.id, hioa.endpoint);
+        }
```

Comment:
Will all these throws trigger in production and kill clusters?

Reply:
They indicate an operator error related to the optional ignore-nodes values.
They won't kill the cluster, but rather kill the replacing node, preventing it from joining, which is what they're supposed to do as the ignore nodes host_id:s must refer to the dead nodes that are no longer in the cluster.
Unfortunately, we still must know their assigned ip address until gossiper transitions to use host_id:s.
In the future we can support host_id only and drop the support for identifying ignore-nodes using ip addresses.

Cc @tgrabiec